src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs

```diff
@@ -95,95 +110,45 @@ public override void Initialize(AnalysisContext analysisContext)
-                }
-                else if (originalDefinition.Equals(gListType))
-                {
-                    allInterfacesStatus.GenericIListPresent = true;
-                }
+                allInterfaces.Add(@interface);
             }
 
-            INamedTypeSymbol? missingInterface;
-            INamedTypeSymbol? implementedInterface;
-            if (allInterfacesStatus.GenericIListPresent)
-            {
-                // Implemented IList<T>, meaning has all 3 generic interfaces. Nothing can be wrong.
-                return;
-            }
-            else if (allInterfacesStatus.IListPresent)
-            {
-                // Implemented IList but not IList<T>.
-                missingInterface = gListType;
-                implementedInterface = iListType;
-            }
-            else if (allInterfacesStatus.GenericICollectionPresent)
+            // First we need to try to match all types from the user definition...
+            if (!userMap.IsEmpty)
             {
-                // Implemented ICollection<T>, and doesn't have an inherit of IList. Nothing can be wrong
-                return;
-            }
-            else if (allInterfacesStatus.ICollectionPresent)
-            {
-                // Implemented ICollection but not ICollection<T>
-                missingInterface = gCollectionType;
-                implementedInterface = iCollectionType;
-            }
-            else if (allInterfacesStatus.GenericIEnumerablePresent)
-            {
-                // Implemented IEnumerable<T>, and doesn't have an inherit of ICollection. Nothing can be wrong
-                return;
+                foreach (var @interface in allInterfaces)
+                {
+                    if (!@interface.IsGenericType &&
+                        userMap.TryGetValue(@interface, out var genericInterface) &&
+                        genericInterface?.IsGenericType == true)
+                    {
+                        ReportDiagnostic(@interface, genericInterface);
+                        return;
+                    }
+                }
             }
-            else if (allInterfacesStatus.IEnumerablePresent)
+
+            // ...Then we can proceed with the hardcoded ones keeping the declaration order
+            for (int i = 0; i < interfacePairs.Length; i++)
             {
-                // Implemented IEnumerable, but not IEnumerable<T>
-                missingInterface = gEnumerableType;
-                implementedInterface = iEnumerableType;
+                var kvp = interfacePairs[i];
+
+                if (allInterfaces.Contains(kvp.Key) && !allInterfaces.Contains(kvp.Value))
+                {
+                    ReportDiagnostic(kvp.Key, kvp.Value);
+                    return;
+                }
             }
-            else
+
+            void ReportDiagnostic(INamedTypeSymbol @interface, INamedTypeSymbol genericInterface)
```

Comment:
Nit: please add an explicit return statement before start of local function definitions.

Reply:
Is it for reading purpose or is there any impact?

Reply:
Yep, just a readability suggestion.